### DISCUSSION TITLE
Triples in Trainer

### DISCUSSION THREAD

**Discussion by seplee** [stanford-futuredata/ColBERT#372] [Q&A]
Hi! 

I'm trying to train answerai-colbert-small-v1 on data retrieved from a recommendation system, in this [sort of scheme](https://arxiv.org/html/2406.11706v1).

Thanks for posting an [example](https://github.com/stanford-futuredata/ColBERT?tab=readme-ov-file#advanced-training-colbertv2-style) on your README.md!

What are triples in `Trainer`? 
I printed out a few characters of the commented MS MARCO file, but I am confused. What do these numbers mean?
<img width="1069" alt="image" src="https://github.com/user-attachments/assets/6c03f117-1a64-420e-a12f-84336d2a624e">

Are they a mapping of the index of one collection to a list of queries and the relevance score of the queries?
ex: `[indexed collection, [indexed query, relevance score] ... ]`?

If so, do I need to [index](https://github.com/stanford-futuredata/ColBERT?tab=readme-ov-file#indexing) my collections and queries in order for Trainer to use them?

Thanks,
Lee

**Comment by robinnarsinghranabhat** [comment]
Each Line represents a query, and how relavant it is to different other documents. 

`[ query_id, (doc_id, doc_score),  (doc_id, doc_score), ...]`

You will have two other `tsv` files which have columns `(query_id, actual_query)` and `(doc_id, actual_document)`.

If you take a look in training datasets like https://huggingface.co/datasets/microsoft/ms_marco , Usually, Query will have a High Relavance associated with one of the Documents, 
And none with Rest. 


Conceptually, I don't see any need to Index for training in colbert's style. And Going through the Training code itself, I can verify, it's not done here
